### PR TITLE
Implement Download of Binary Resources

### LINKED ISSUE #2108: Implement Direct Download of Binary Content at the Binary Endpoint
Scaffolding:

 * in `blaze.rest-api.routes` introduce a new `:response-type` called `:binary`
   * use that new response type in `wrap-output` to call a new middleware function called `blaze.middleware.fhir.output/wrap-binary-output`
   * in `resource-route` tag the route with  `:response-type` `:binary` if the `name` is "Binary"
 * create a new middleware function called `blaze.middleware.fhir.output/wrap-binary-output`
   * it should call `request-format` like `wrap-output` but only handle `:fhir+json` and `:fhir+xml` in the normal way and return the binary content otherwise
 * the binary handling should take the `:data` base64 string of the Binary resource, base64 decode it and output it as body and also set the `:contentType` string as `Content-Type` header

For tests:
 * unit tests in `blaze.middleware.fhir.output-test`
 * integration test in GitHub Actions by creating a new Bash script in `.github/scripts`
   * `generate-large-binary-resource.sh`
   * `create-patient.sh` 

Part of #2009.

### PR DESCRIPTION
**Pull request by allentiak**
Fixes #2108

### CODE REVIEW

File: docs/implementation/fhir-data-model.md [hunk 1]
```diff
@@ -14,6 +14,7 @@ Parsing a JSON document like:
 {
  "resourceType": "Patient",
  "id": "0",
+ "name": "John Doe",
```
Comment (alexanderkiel) [inline docs/implementation/fhir-data-model.md:17]: name isn't a string. please check the FHIR spec how a HumanName will look

File: modules/rest-api/src/blaze/rest_api/routes.clj [hunk 2]
```diff
@@ -97,127 +98,129 @@
...
-       (cond-> {:name (keyword name "type")}
-         (contains? interactions :search-type)
-         (assoc :get {:interaction "search-type"
-                      :middleware [[wrap-db node db-sync-timeout]
+   (cond->
+    [(str "/" name)
+     {:fhir.resource/type name}
+     [""
+      (cond-> {:name (keyword name "type")}
+        (= name "Binary")
```
Comment (alexanderkiel) [inline modules/rest-api/src/blaze/rest_api/routes.clj:106]: this is perfect

File: modules/rest-api/src/blaze/rest_api/routes.clj [hunk 3]
```diff
@@ -97,127 +98,129 @@
...
+        (assoc :response-type :binary)
+        (contains? interactions :search-type)
+        (assoc :get {:interaction "search-type"
+                     :middleware [[wrap-db node db-sync-timeout]
+                                  wrap-link-headers]
```
Comment (alexanderkiel) [inline modules/rest-api/src/blaze/rest_api/routes.clj:111]: I assume that you did not rebase properly onto the current main. thats why you have so much change here.

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 4]
```diff
@@ -115,6 +116,14 @@
       (s2/unform spec resource)
       (throw (ex-info (format "Missing spec: %s" key) {:key key})))))
 
+(defn unform-binary
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:119]: I would not create an unform binary at all. Unforming is the transformation from the internal format into JSON or XML, which are both structured formats of the whole resource. Outputting the binary part of the binary resource is just an extraction of a single value from the internal format.

File: modules/rest-util/src/blaze/middleware/fhir/output.clj [hunk 5]
```diff
@@ -1,5 +1,7 @@
 (ns blaze.middleware.fhir.output
-  "JSON/XML serialization middleware."
+  "Resources serialization middleware.
+
+  Currently supports JSON, XML, and binary."
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/middleware/fhir/output.clj:4]: "binary" isn't a common resource serialization format like JSON or XML. It's rather a special serialization of the Binary resource only. I would write: 

FHIR JSON/XML serialization middleware.

In addition to the common JSOM/XML serializations, `wrap-binary-output` will
output the binary data of the Binary resource if requested.

File: modules/rest-util/src/blaze/middleware/fhir/output.clj [hunk 6]
```diff
@@ -45,6 +47,11 @@
   (with-open [_ (prom/timer generate-duration-seconds "xml")]
     (generate-xml* body)))
 
+(defn- generate-binary [body]
+  (log/trace "generate binary")
+  (with-open [_ (prom/timer generate-duration-seconds "binary")]
+    (fhir-spec/unform-binary body)))
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/middleware/fhir/output.clj:53]: Instead of the unform-binary, I would just take the value of the data property and base64 decode it.

File: modules/rest-util/src/blaze/middleware/fhir/output.clj [hunk 7]
```diff
@@ -53,6 +60,10 @@
   (cond-> response body (-> (update :body generate-xml)
                             (ring/content-type content-type))))
 
+(defn- encode-response-binary [{:keys [body] :as response} content-type]
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/middleware/fhir/output.clj:63]: we have to take the content-type from the Binary resource

File: modules/rest-util/src/blaze/middleware/fhir/resource.clj [hunk 8]
```diff
@@ -1,5 +1,7 @@
 (ns blaze.middleware.fhir.resource
-  "JSON/XML deserialization middleware."
+  "Resource deserialization middleware.
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/middleware/fhir/resource.clj:2]: please leave it, it's not part of that PR

File: modules/rest-util/src/blaze/middleware/output.clj [hunk 9]
```diff
@@ -1,5 +1,7 @@
 (ns blaze.middleware.output
-  "JSON serialization middleware."
+  "Data serialization middleware.
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/middleware/output.clj:2]: please leave it like it was

**Review by alexanderkiel — CHANGES_REQUESTED** [review]

File: modules/fhir-structure/src/blaze/fhir/spec.clj (hunk 4 above)
Comment (allentiak) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:119]: I'll remove this last level of indirection.

File: modules/rest-util/src/blaze/middleware/fhir/output.clj (hunk 7 above)
Comment (allentiak) [reply modules/rest-util/src/blaze/middleware/fhir/output.clj:63]: https://www.hl7.org/fhir/R4/binary.html

File: modules/rest-util/src/blaze/middleware/fhir/output.clj [hunk 10]
```diff
@@ -92,3 +103,17 @@
   ([handler opts]
    (fn [request respond raise]
      (handler request #(respond (handle-response opts request %)) raise))))
+
+(defn handle-binary-response [opts request response]
+  (case (request-format request)
+    :fhir+json (encode-response-json response "application/fhir+json;charset=utf-8")
+    :fhir+xml (encode-response-xml response "application/fhir+xml;charset=utf-8")
+    (encode-response-binary response "text/plain;charset=utf-8")))
```
Comment (allentiak) [inline modules/rest-util/src/blaze/middleware/fhir/output.clj:111]: Delete the string from the parameters list.
See line 65.

File: modules/rest-util/src/blaze/middleware/fhir/resource.clj (hunk 8 above)
Comment (allentiak) [reply modules/rest-util/src/blaze/middleware/fhir/resource.clj:2]: FHIR Resource deserialization middleware.
...

File: modules/rest-util/src/blaze/middleware/output.clj (hunk 9 above)
Comment (allentiak) [reply modules/rest-util/src/blaze/middleware/output.clj:2]: "Non-FHIR JSON deserialization middleware."

File: profiling/blaze/profiling.clj
Comment (allentiak) [inline profiling/blaze/profiling.clj:1]: Add this file to `cljfmt` target(s).

Comment (allentiak) [reply profiling/blaze/profiling.clj:1]: Maintain the same style in `/dev/blaze/dev.clj`.
(For the future, find a way to scan all files with `parinfer`.)

File: modules/rest-util/src/blaze/middleware/fhir/output.clj (hunk 5 above)
Comment (allentiak) [reply modules/rest-util/src/blaze/middleware/fhir/output.clj:4]: I will move all implementation-specific mentions to the specific function's clojuredoc string, so we don't have to keep changing it here.

File: docs/implementation/fhir-data-model.md (hunk 1 above)
Comment (allentiak) [reply docs/implementation/fhir-data-model.md:17]: No longer necessary after 0.30.1 release.

Comment (allentiak) [reply docs/implementation/fhir-data-model.md:17]: https://hl7.org/fhir/R4B/datatypes.html#HumanName

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 11]
```diff
@@ -205,5 +211,12 @@
       [:headers "Content-Type"] := "application/fhir+xml;charset=utf-8"
       [:body parse-xml :issue 0 :diagnostics] := "Invalid white space character (0x1e) in text to output (in xml 1.1, could output as a character entity)")))
 
+(deftest binary-test
+  (testing "with accept header"
+    (given (call binary-resource-handler-200 {:headers {"accept" "text/plain"}})
+      :status := 200
+      [:headers "Content-Type"] := "text/plain"
```
[the remark this answers is not in the document]
Comment (allentiak) [reply modules/rest-util/test/blaze/middleware/fhir/output_test.clj:218]: I will add more tests once I have made this one pass.

File: docs/implementation/fhir-data-model.md [hunk 12]
```diff
@@ -14,6 +14,9 @@ Parsing a JSON document like:
 {
  "resourceType": "Patient",
  "id": "0",
+ "name": {
```
Comment (alexanderkiel) [inline docs/implementation/fhir-data-model.md:17]: Name should be an array of objects.

File: modules/rest-util/src/blaze/middleware/fihr.clj [hunk 13]
```diff
@@ -0,0 +1,2 @@
+(ns blaze.middleware.fhir
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/middleware/fihr.clj:1]: Should be removed.

File: modules/rest-util/src/blaze/middleware/fhir/output.clj
Comment (allentiak) [inline modules/rest-util/src/blaze/middleware/fhir/output.clj:1]: Somehow, it seems this file was changed more than needed by accident. I will remove the unrelated changes.

**Review by allentiak — COMMENTED** [review]
Just re-reviewed my own changes.

File: profiling/blaze/profiling.clj
Comment (allentiak) [reply profiling/blaze/profiling.clj:1]: Moved to #2125.

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 14]
```diff
@@ -205,5 +220,26 @@
...
+      [:body bs/from-byte-array] := #blaze/byte-string"3130353631340A"))
+  (testing "without data and with content type"
+    (given (call (binary-resource-handler-200 {:content-type "text/plain" :data nil}) {:headers {"accept" "text/plain"}})
+      :status := 200
+      [:headers "Content-Type"] := "text/plain"
+      [:body := nil]))
+  (testing "without body with content type"
+    (given (call (binary-resource-handler-200 nil) {:headers {"accept" "text/plain"}})
+      :status := 200
```
Comment (allentiak) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:241]: I am not sure whether this last test is correct...

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 15]
```diff
@@ -205,5 +220,26 @@
...
       [:body parse-xml :issue 0 :diagnostics] := "Invalid white space character (0x1e) in text to output (in xml 1.1, could output as a character entity)")))
 
+(deftest binary-test
+  (testing "with data and without content type"
+    (given (call (binary-resource-handler-200 {:content-type nil :data #fhir/base64Binary"MTA1NjE0Cg=="}) {:headers {"accept" "text/plain"}})
+      :status := 200
+      [:headers "Content-Type"] := "application/octet-stream"
+      [:body bs/from-byte-array] := #blaze/byte-string"3130353631340A"))
+  (testing "with data and with content type"
+    (given (call (binary-resource-handler-200 {:content-type "text/plain" :data #fhir/base64Binary"MTA1NjE0Cg=="}) {:headers {"accept" "text/plain"}})
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:230]: either the content-type should be a FHIR code or the data should not be a FHIR base64Binary in order to keep the arguments uniform

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 16]
```diff
@@ -205,5 +220,26 @@
       [:headers "Content-Type"] := "application/fhir+xml;charset=utf-8"
       [:body parse-xml :issue 0 :diagnostics] := "Invalid white space character (0x1e) in text to output (in xml 1.1, could output as a character entity)")))
 
+(deftest binary-test
+  (testing "with data and without content type"
+    (given (call (binary-resource-handler-200 {:content-type nil :data #fhir/base64Binary"MTA1NjE0Cg=="}) {:headers {"accept" "text/plain"}})
+      :status := 200
+      [:headers "Content-Type"] := "application/octet-stream"
+      [:body bs/from-byte-array] := #blaze/byte-string"3130353631340A"))
+  (testing "with data and with content type"
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:229]: please add a newline between testing blocks to improve readability

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 17]
```diff
@@ -205,5 +220,26 @@
...
+    (given (call (binary-resource-handler-200 {:content-type "text/plain" :data #fhir/base64Binary"MTA1NjE0Cg=="}) {:headers {"accept" "text/plain"}})
+      :status := 200
+      [:headers "Content-Type"] := "text/plain"
+      [:body bs/from-byte-array] := #blaze/byte-string"3130353631340A"))
+  (testing "without data and with content type"
+    (given (call (binary-resource-handler-200 {:content-type "text/plain" :data nil}) {:headers {"accept" "text/plain"}})
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:235]: just don't specify data at all

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 18]
```diff
@@ -205,5 +220,26 @@
...
+      :status := 200
+      [:headers "Content-Type"] := "text/plain"
+      [:body := nil]))
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:238]: `:= nil` has to be outside of the path vector. in this case you don't need the vector at all like with status

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj (hunk 14 above)
Comment (alexanderkiel) [reply modules/rest-util/test/blaze/middleware/fhir/output_test.clj:241]: it's also without content type und I would add the test of body being nil

**Review by alexanderkiel — CHANGES_REQUESTED** [review]

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj (hunk 18 above)
Comment (allentiak) [reply modules/rest-util/test/blaze/middleware/fhir/output_test.clj:238]: `:body := nil`, then?

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj (hunk 14 above)
Comment (allentiak) [reply modules/rest-util/test/blaze/middleware/fhir/output_test.clj:241]: I changed the parameter of `binary-resource-handler-200 from being the `content-type` to being the whole `body`.

**Review by allentiak — COMMENTED** [review]
Known issues:
1. I still need to add the integration tests.
2. There are some formatting/linting errors. However, a local run with the latest versions of both `clj-kondo` (2024.09.27) and `cljfmt` (0.13.0 - 2024.09.28) don't give me any error (see below)...
It seems I will have to fix this manually?
```
[modules/rest-util] (binary-upload*)
% clj-kondo --lint src test deps.edn
linting took 200ms, errors: 0, warnings: 0
-> % cljfmt check src test deps.edn tests.edn
All source files formatted correctly
```

The rest is ready for review.

**Review by allentiak — COMMENTED** [review]
_(Just to mark the files as viewed.)_

File: modules/rest-util/src/blaze/middleware/fhir/output.clj [hunk 19]
```diff
@@ -53,6 +66,14 @@
   (cond-> response body (-> (update :body generate-xml)
                             (ring/content-type content-type))))
 
+(defn- encode-response-binary [{:keys [body] :as response}]
+  (let [content-type (or (-> response :body :contentType fhir-type/value)
+                         "application/octet-stream")]
+    (cond->
+        (ring/content-type response content-type)
+      body (-> (update :body generate-binary)))))
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/middleware/fhir/output.clj:74]: In case the body is nil, I would also not set the content-type to `application/octet-stream`. it has no value to do it for an empty content. So you can go for the same style as in `encode-response-json` here. Please also write a separate function that takes the body and returns the content-type named `binary-content-type`.

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 20]
```diff
@@ -27,6 +29,19 @@
...
 
+(defn binary-resource-handler-200
+  "A handler which uses the binary middleware and just returns
+  a binary resource."
+  [{:keys [content-type data] :as _body}]
+  (wrap-binary-output
+   (fn [_ respond _]
+     (respond
+       (ring/response
+         (cond->
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:40]: also no need for a newline here

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 21]
```diff
@@ -205,5 +220,37 @@
...
+      [:headers "Content-Type"] := "text/plain"
+      :body := nil))
+
+  (testing "without data and without content type"
+    (given (call (binary-resource-handler-200 {:content-type nil}) {:headers {"accept" "text/plain"}})
+      :status := 200
+      [:headers "Content-Type"] := "application/octet-stream"
+      :body := nil))
+
+  (testing "without body at all"
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:249]: this test will create a body, because you always create a binary resource. I would use the `resource-handler-304` instead. it creates a 304 response without a body.

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 22]
```diff
@@ -205,5 +220,37 @@
...
+    (given (call (binary-resource-handler-200 nil) {:headers {"accept" "text/plain"}})
+      :status := 200
+      [:headers "Content-Type"] := "application/octet-stream"
+      :body := nil)))
+
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:254]: we also need two additional tests, one that if `application/fhir+json` is requested, that the binary resource in JSON is returned and the other one for XML. please copy the asserts from the corresponding JSON/XML tests

**Review by alexanderkiel — CHANGES_REQUESTED** [review]

File: modules/rest-api/src/blaze/rest_api/routes.clj [hunk 23]
```diff
@@ -64,6 +64,7 @@
    :compile (fn [{:keys [response-type] :response-type.json/keys [opts]} _]
               (condp = response-type
                 :json (output/wrap-output opts)
+                :binary (fhir-output/wrap-binary-output opts)
```
Comment (alexanderkiel) [inline modules/rest-api/src/blaze/rest_api/routes.clj:67]: here, you have to return a middleware function (a function taking a handler). It would be possible to write `#(fhir-output/wrap-binary-output % opts)` as this creates a function taking a handler. but as we don't need the opts, we can just return `fhir-output/wrap-binary-output`.

File: modules/rest-util/src/blaze/middleware/fhir/output.clj [hunk 24]
```diff
@@ -92,3 +113,17 @@
...
    (fn [request respond raise]
      (handler request #(respond (handle-response opts request %)) raise))))
+
+(defn handle-binary-response [_opts request response]
+  (case (request-format request)
+    :fhir+json (encode-response-json response "application/fhir+json;charset=utf-8")
+    :fhir+xml (encode-response-xml response "application/fhir+xml;charset=utf-8")
+    (encode-response-binary response)))
+
+(defn wrap-binary-output
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/middleware/fhir/output.clj:123]: as we don't need the `opts` please remove the arity taking opts

File: modules/rest-util/src/blaze/middleware/fhir/output.clj [hunk 25]
```diff
@@ -53,6 +65,15 @@
   (cond-> response body (-> (update :body generate-xml)
                             (ring/content-type content-type))))
 
+(defn- binary-content-type [body]
+  (when body
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/middleware/fhir/output.clj:69]: the body isn't nil here, because you check for nil already in  `encode-response-binary`. even than both the function `:contentType` and `fhir-type/value` will accept nil and return nil

File: modules/rest-util/src/blaze/middleware/fhir/output.clj [hunk 26]
```diff
@@ -1,8 +1,13 @@
...
-  "JSON/XML serialization middleware."
+  "FHIR Resource serialization middleware.
+
+  Currently supported formats:
+  * standard: JSON, XML;
+  * special: binary."
   (:require
    [blaze.anomaly :as ba]
    [blaze.fhir.spec :as fhir-spec]
+   [blaze.fhir.spec.type :as fhir-type]
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/middleware/fhir/output.clj:10]: please use the alias `type` here as every other require uses. If you are unsure, just search in the whole project for `blaze.fhir.spec.type :as` and you will the the common alias

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 27]
```diff
@@ -205,5 +227,47 @@
       [:headers "Content-Type"] := "application/fhir+xml;charset=utf-8"
       [:body parse-xml :issue 0 :diagnostics] := "Invalid white space character (0x1e) in text to output (in xml 1.1, could output as a character entity)")))
 
+(deftest binary-resource-test
+  (testing "possible accept headers"
+    (are [accept content-type body]
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:232]: please don't combine `given` with `are` because `given` already uses `is` inside. it's better to use `doseq` together with `given`

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 28]
```diff
@@ -205,5 +227,47 @@
...
+         (given (call (binary-resource-handler-200 {:content-type "text/plain" :data "MTA1NjE0Cg=="}) {:headers {"accept" accept}})
+           :status := 200
+           [:headers "Content-Type"] := content-type
+           [:body bs/from-byte-array] := body)
+      "application/fhir+json" "application/fhir+json;charset=utf-8" #blaze/byte-string"7B2264617461223A224D5441314E6A453043673D3D222C22636F6E74656E7454797065223A22746578742F706C61696E222C227265736F7572636554797065223A2242696E617279227D"
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:237]: I would not byte-compare the FHIR output. please use `parse-json` and `parse-xml` like in the other tests

Comment (alexanderkiel) [reply modules/rest-util/test/blaze/middleware/fhir/output_test.clj:237]: saying that, please just write two `testing` blocks, one for JSON and one for XML

**Review by alexanderkiel — CHANGES_REQUESTED** [review]

File: modules/rest-util/src/blaze/middleware/fhir/output.clj (hunk 25 above)
Comment (allentiak) [reply modules/rest-util/src/blaze/middleware/fhir/output.clj:69]: I was just being unnecessarily defensive here, ha ha.

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj (hunk 27 above)
Comment (allentiak) [reply modules/rest-util/test/blaze/middleware/fhir/output_test.clj:232]: Should we also rewrite the existing tests that use `are` to use `is`?

File: .github/scripts/read-binary-content.sh [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+echo "Created Binary resource that contains the Random Data, with ID: $ID"
+
+# Retrieve the Binary resource, and Base64 encode it so it can be safely handled by Bash
+BASE64_ENCODED_BINARY_RESOURCE=$(curl -s -H 'Accept: application/pdf' "$BASE/Binary/$ID" | base64 | tr -d '\n')
+
+echo "Binary Result, Base 64 encoded: $BASE64_ENCODED_BINARY_RESOURCE"
+
+echo "Binary data retrieved. Verifying content..."
+
+test "Base64 encoding of Resource data" "$DATA" "$BASE64_ENCODED_BINARY_RESOURCE"
```
Comment (alexanderkiel) [inline .github/scripts/read-binary-content.sh:38]: Please add a test reading the Binary resource using application/fhir+json and test that the base64 content is the same as the original content.

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 30]
```diff
@@ -21,22 +23,42 @@
...
   "A handler which just returns a patient."
   (wrap-output
    (fn [_ respond _]
      (respond (ring/response {:fhir/type :fhir/Patient :id "0"})))))
 
-(def resource-handler-304
+(defn- binary-resource-handler-200
+  "A handler which uses the binary middleware and
+  returns a binary resource."
+  [{:keys [content-type data] :as _body}]
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:35]: If you don't need the body, just don't add the `:as` part.

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 31]
```diff
@@ -56,154 +78,143 @@
...
 
   (testing "failing XML emit"
     (given (call (special-resource-handler {:fhir/type :fhir/Patient :id "0" :gender #fhir/code"foo\u001Ebar"}) {:headers {"accept" "application/fhir+xml"}})
+      :status := 500
       [:headers "Content-Type"] := "application/fhir+xml;charset=utf-8"
+      [:body parse-xml :fhir/type] := :fhir/OperationOutcome
       [:body parse-xml :issue 0 :diagnostics] := "Invalid white space character (0x1e) in text to output (in xml 1.1, could output as a character entity)")))
 
+(deftest binary-resource-test
+
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:169]: I usually don't use a newline after `deftest`. I use it only to separate the `testing` blocks.

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 32]
```diff
@@ -56,154 +78,143 @@
...
+  (testing "possible accept headers"
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:170]: I would say "returning the resource". "possible accept headers" is just a copy from above. But here we have to differentiate between returning the resource and returning the data.

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 33]
```diff
@@ -56,154 +78,143 @@
...
+
+    (testing "XML"
+      (given (call (binary-resource-handler-200 {:content-type "text/plain" :data "MTA1NjE0Cg=="}) {:headers {"accept" "application/fhir+xml"}})
+        :status := 200
+        [:headers "Content-Type"] := "application/fhir+xml;charset=utf-8"
+        [:body parse-xml] := {:fhir/type :fhir/Binary
+                              :contentType #fhir/code"text/plain"
+                              :data #fhir/base64Binary"MTA1NjE0Cg=="})))
+
+  (testing "with data"
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:187]: I would say: "returning the data" here

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 34]
```diff
@@ -21,22 +23,42 @@
...
+  (wrap-binary-output
+   (fn [_ respond _]
+     (respond (ring/status 200)))))
+
 (defn- special-resource-handler [resource]
   (wrap-output
    (fn [_ respond _]
-     (respond (ring/response resource)))))
+     (respond (-> (ring/response resource)
+                  (ring/status 500))))))
```
Comment (alexanderkiel) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:61]: The Internal Server Error hat to come from the parsing error and not from the handler above the middleware chain.

**Review by alexanderkiel — CHANGES_REQUESTED** [review]

File: modules/rest-util/src/blaze/middleware/output.clj [hunk 35]
```diff
@@ -1,5 +1,8 @@
 (ns blaze.middleware.output
-  "JSON serialization middleware."
+  "Non-FHIR serialization middleware.
```
Comment (allentiak) [inline modules/rest-util/src/blaze/middleware/output.clj:2]: BTW: @alexanderkiel would you have any suggestion for a better name for this?
"Non-FHIR" seems too general...

Comment (alexanderkiel) [reply modules/rest-util/src/blaze/middleware/output.clj:2]: Non-FHIR is ok. I use this middleware for the Admin-API were I return Non-FHIR JSON.

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj (hunk 34 above)
Comment (alexanderkiel) [reply modules/rest-util/test/blaze/middleware/fhir/output_test.clj:61]: The code handling the XML generating error has to set the status to 500. In case the base64 encoding can throw an error, you should do the same.

Comment (allentiak) [reply modules/rest-util/test/blaze/middleware/fhir/output_test.clj:61]: Then, I should add the corresponding regression tests, both for binary and non-binary resources, both for XML and JSON?

Comment (alexanderkiel) [reply modules/rest-util/test/blaze/middleware/fhir/output_test.clj:61]: Yes please add the tests for JSON, XML and Binary.

File: modules/rest-util/test/blaze/middleware/fhir/output_test.clj [hunk 36]
```diff
@@ -21,22 +23,48 @@
...
 
-(def resource-handler-200
+(def ^:private resource-handler-200
   "A handler which just returns a patient."
   (wrap-output
    (fn [_ respond _]
      (respond (ring/response {:fhir/type :fhir/Patient :id "0"})))))
 
-(def resource-handler-304
+(defn- binary-resource-handler-200
```
Comment (allentiak) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:32]: Here

Comment (allentiak) [inline modules/rest-util/test/blaze/middleware/fhir/output_test.clj:32]: Here.

**Review by alexanderkiel — APPROVED** [review]
